File: arthive-backend/app/graphql/types/review_type.rb
module Types
    class ReviewType < Types::BaseObject
        field :id, ID, null: false
        field :content, String, null: true
        field :rating, Float, null: true
        field :if_favorite, Boolean, null: false
        field :if_finished, Boolean, null: false
        field :created_at, GraphQL::Types::ISO8601DateTime, null: false
        field :updated_at, GraphQL::Types::ISO8601DateTime, null: false

        field :user, Types::UserType, null: false
        field :media, Types::MediaType, null: false

        field :review_comments, [Types::ReviewCommentType], null: true
        field :review_likes, [Types::ReviewLikeType], null: true

        field :like_count, Integer, null: false
        field :comment_count, Integer, null: false

        field :if_liked, Boolean, null: false

        field :image_details, [Types::ImageDetailsType], null: true

        def image_details
            object.images_blobs.map do |blob|
                {
                    signed_id: blob.signed_id,
                    url: blob.url(expires_in: 1.hour)
                }
            end
        end

        def if_liked
            object.review_likes.exists?(user_id: context[:current_user].id)
        end

        def like_count
            object.review_likes.count
        end

        def comment_count
            object.review_comments.count
        end
        
    end
end

File: arthive-frontend/src/types/review_type.ts
export type UserReview = {
    id: number
    content?: string
    rating?: number
    ifFavorite: boolean
    ifFinished: boolean
    imageDetails: {
        signedId: string
        url: string
    }[]
}

type ImageDetail = {
    signedId: 
    string
    url: string
}

export type Review = {
    id: number
    content?: string
    rating?: number
    ifFavorite: boolean
    ifFinished: boolean
    updatedAt: string

    user: {
        id: number
        username: string
        profilePicture: string | null
    }

    likeCount: number,
    commentCount: number,
    ifLiked: boolean,
    imageDetails: ImageDetail[]
    media: {
        id: number
        title: string
        coverImage: string
        creator: string
        year: number
        genre: string[]
        contentType: "book" | "film" | "series" | "game"
    }
}

export type AllReview = {
    id: number
    content?: string
    rating?: number
    ifFavorite: boolean
    ifFinished: boolean
    updatedAt: string
    media: {
        id: number
        title: string
        coverImage: string
        creator: string
        year: number
        genre: string[]
        contentType: "book" | "film" | "series" | "game"
    }

    imageDetails: ImageDetail[]
    likeCount: number,
    commentCount: number,
    ifLiked: boolean,
}


export type MainReview = {
    id: number
    content?: string
    rating?: number
    ifFavorite: boolean
    ifFinished: boolean
    updatedAt: string

    user: {
        id: number
        username: string
        profilePicture: string | null
    }
    media: {
        id: number
        title: string
        coverImage: string
        creator: string
        year: number
        genre: string[]
        contentType: "book" | "film" | "series" | "game"
    }
    imageDetails: ImageDetail[]
    likeCount: number
    commentCount: number
    ifLiked: boolean
}

export type ReviewComment = {
    id: number
    comment: string
    createdAt: string

    user: {
        id: number
        username: string
        profilePicture: string | null
    }
}

export type ReviewPage = {
    review: MainReview
    reviewComments: {
        edges: {
            node: ReviewComment
        }[]
        pageInfo: {
            hasNextPage: boolean
            endCursor: string | null
        }
    }
}